### ISSUE TITLE
nit / design question: unnecessary coupling?

### ISSUE THREAD

**Issue by bennetyee** [oasisprotocol/oasis-bridge#70]
https://github.com/oasisprotocol/oasis-bridge/blob/80f5189ca339063c644ff5dce3431ba4f35062f7/module-bridge/src/lib.rs#L263

we require `tx_witness` here and below at

https://github.com/oasisprotocol/oasis-bridge/blob/80f5189ca339063c644ff5dce3431ba4f35062f7/module-bridge/src/lib.rs#L316

for `tx_release` that the caller is an authorized witness.  since witness entries are signed statements, why not allow anybody to submit the gathered signatures?  that way, witnesses can be more obviously implemented as two decoupled processes -- an observer of events that generate signed witness statements, and a submitter of these witness statements.  depending on calldata limits, we might be able to save some overhead if a single tx can submit several signed witness statements, though maybe that's excess complexity and the savings is not large enough?

one hopes that the witnesses would use a decoupled design so that if the oasis transaction submission mechanism is down -- or network partitioned briefly -- it'd be easier to recover.  the remote-blockchain observer / signature generation component of the witness can obviously run asynchrously ahead of the oasis-transaction-submitter component.  so even if we want to have access limited to `caller_address`, we could have the signing key for the witness statement be distinct from the transaction submission signing key.  having distinct keys would also eliminate concerns with potential message parsing ambiguities where a transaction (or something else) might look like a witness or release events being signed, since distinct keys would be single-use.

**Comment by pro-wh** [comment]
> why not allow anybody to submit the gathered signatures

that's somewhat allowed, if witnesses would publish their signed transactions out of band. anyone can submit it for them. although there is the issue of the nonces having to be right.

> if the oasis transaction submission mechanism is down -- or network partitioned briefly -- it'd be easier to recover

`tx_release` is for minting/unlocking tokens on the oasis side, so if the oasis network were down, there's nothing else to do even if you did verify the signatures

**Comment by bennetyee** [comment]
as i read it, only a witness can call `tx_witness` and `tx_release`.  since the signature verification code isn't written yet, it's not totally clear, but the check for duplicate signatures is checking using `index` from the `.find(...)` check, so it looks like the intent is that only the signer of the witness statement can call the transaction (rather than, say, any witness).

what i was thinking re decoupling is that anybody (rather than the signing witness or even any other witness) can deliver the signed witness statement.  regarding what `tx_release` does -- what i was saying is that in either case (whether Ethereum side somebody locked Eth [or other ERC20 tokens] or burned eROSE), the witness component that observes Ethereum events don't have to block in a decoupled design, and can generate signed witness statements to be queued up internally (or even broadcast) while waiting for Oasis submission to be possible again.  no signature verification would occur, since the observer just signed, and the bridge module (which is inaccessible) is the one that would do the signature verification.  assuming that Oasis submission and transaction processing is faster than Ethereum observation, this would allow the Oasis pier (bridge module) to catch up faster than a tightly-coupled, synchronous witness design.

the broadcast instead of queuing up internally case is just the design variation where anybody can submit witness statements on the observer's behalf.  presumably if we had SLA requirements for the witnesses, the witness operator will want to also operate a oasis-transaction-submitter component, so that meeting SLAs won't depend on the kindness of strangers.

**Comment by pro-wh** [comment]
there's this implicit 'anyone can submit' feature that comes from the overall design of oasis paratime transactions. the witness signs the transaction resulting in a byte string that's ready to submit. but there's no additional authentication during the actual submission, which allows for this sort of decoupled sign -- submit workflow.

it should be similar to how oasis's consensus layer supports signing transactions on an airgapped machine and having an internet-connected machine submit that transaction.

**Comment by kostko** [comment]
Yeah the difference is whether it is the transactions that are authenticated or the cross-chain messages (or both). I agree that it would make sense that anyone could submit the signed statement and that this would result in less coupling. Especially in any more generalized design.
